In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns
sns.set()

In [2]:
raw_data = pd.read_csv('2.02. Binary predictors.csv')
raw_data.head()

SAT Admitted  Gender
0  1363       No    Male
1  1792      Yes  Female
2  1954      Yes  Female
3  1653       No    Male
4  1593       No    Male

In [3]:
data = raw_data.copy()
data['Admitted'] = data['Admitted'].map({'Yes':1,'No':0})
data['Gender'] = data['Gender'].map({'Female':1,'Male':0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [5]:
y=data['Admitted']
x1=data[['SAT','Gender']]

In [6]:
x = sm.add_constant(x1)
reg = sm.Logit(y,x).fit()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


In [7]:
reg.pred_table()

array([[69.,  5.],
       [ 4., 90.]])

In [8]:
#cm_df = confusion matrix data frame
cm_df = pd.DataFrame(reg.pred_table())
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df.index = ['Actual 0','Actual 1']
cm_df

Predicted 0  Predicted 1
Actual 0         69.0          5.0
Actual 1          4.0         90.0

In [9]:
cm = np.array(cm_df)
accuracy_train = (cm[0,0]+cm[1,1])/cm.sum()
accuracy_train

0.9464285714285714

In [10]:
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Sat, 07 Aug 2021   Pseudo R-squ.:                  0.8249
Time:                        15:35:27   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

### Interpretation of Binary Predictors coef.

In [20]:
f = 1
m = 0
result = (np.exp(1.945)*(f-m)).round(2)
print('The odds of female to get admitted is {} times greater than the odds of males to get admitted in this dataset'.format(result))

The odds of female to get admitted is 6.99 times greater than the odds of males to get admitted in this dataset


### How to predict

In [18]:
SAT= int(input())
Gender = int(input())
yhat = reg.params[0]+reg.params[1]*SAT+reg.params[2]*Gender
p = 1/(1+np.exp(-yhat))
if p>0.5:
    print('Probability:',(p*100))
    print('1')
else:
    print('Probability:',p*100)
    print('0')
    

1750
0
Probability: 93.72803533680644
1


In [19]:
99-93.7

5.299999999999997

### For the same SAT number 1750,
#### Probability of female to get admitted is = 99
#### Probability of male to get admitted is = 93.7
#### 99-93.7 = 5.3 times greater for female (almost accurate)